src/utils/xhrRetryHook.ts

```diff
@@ -0,0 +1,123 @@
+import retry from "retry"
+
+export interface RetryOptions {
+  retries: number
+  factor: number
+  minTimeout: number
+  maxTimeout: number
+  randomize: boolean
+  retryableStatusCodes: number[]
+}
+
+export interface DWCRequestHookMetadata {
+  url: string
+  method: string
+}
+
+const defaultRetryOptions = {
+  retries: 5,
+  factor: 3,
+  minTimeout: 1 * 1000,
+  maxTimeout: 60 * 1000,
+  randomize: true,
+  retryableStatusCodes: [429, 500],
+};
+
+let retryOptions = { ...defaultRetryOptions }
+
+/**
+ * Request hook used to add retry functionality to XHR requests.
+ *
+ * @param {XMLHttpRequest} request XHR request instance
+ * @param {object} metadata Metadata about the request
+ * @param {object} metadata.url URL
+ * @param {object} metadata.method HTTP method
+ * @returns {XMLHttpRequest} request instance optionally modified
+ */
+const xhrRetryHook = (request: XMLHttpRequest, metadata: DWCRequestHookMetadata) => {
+  const { url, method } = metadata
+
+  function faultTolerantRequestSend(...args: any) {
+    const operation = retry.operation(retryOptions)
+
+    operation.attempt(function operationAttempt(currentAttempt) {
+      const noop = () => {}
+      const originalOnReadyStateChange = request.onreadystatechange || noop
+
+      /** Overriding/extending XHR function */
+      request.onreadystatechange = function onReadyStateChange(...args: any) {
+        originalOnReadyStateChange.apply(request, args);
+
+        if (retryOptions.retryableStatusCodes.includes(request.status)) {
+          const errorMessage = `Attempt to request ${url} failed.`
+          const attemptFailedError = new Error(errorMessage)
+          operation.retry(attemptFailedError)
+        }
+      }
+
+      /** Call open only on retry (after headers and other things were set in the xhr instance) */
+      if (currentAttempt > 1) {
+        console.warn(`Requesting ${url}... (attempt: ${currentAttempt})`)
+        request.open(method, url, true)
+      }
+    })
+
+    originalRequestSend.apply(request, args)
+  }
+
+  /** Overriding/extending XHR function */
+  const originalRequestSend = request.send;
+  request.send = faultTolerantRequestSend;
+
+  return request;
+};
+
+/**
+ * Returns a configured retry request hook function
+ * that can be used to add retry functionality to XHR request.
+ *
+ * Default options:
+ *   retries: 5
+ *   factor: 3
+ *   minTimeout: 1 * 1000
+ *   maxTimeout: 60 * 1000
+ *   randomize: true
+ *
+ * @param {object} options
+ * @param {number} options.retires number of retries
+ * @param {number} options.factor factor
+ * @param {number} options.minTimeout the min timeout
+ * @param {number} options.maxTimeout the max timeout
+ * @param {boolean} options.randomize randomize
+ * @param {array} options.retryableStatusCodes status codes that can trigger retry
+ * @returns {function} the configured retry request function
```

Comment:
Two things:

* I am trying to standardize docstrings to use the `@param options.name - Description` pattern:
  - no types to void typos (types are already annotated via Typescript)
  - a `-` in between parameter name and description
  - Capitalize first word of description
* Some of the parameters are really hard to understand and the description could be further clarified 

```suggestion
 * @param options
 * @param options.retries - Number of retries
 * @param options.factor - Factor
 * @param options.minTimeout - Min number of seconds to wait before next retry
 * @param options.maxTimeout - Max number of seconds to wait before next retry
 * @param options.randomize - Whether randomization should be applied
 * @param options.retryableStatusCodes - HTTP status codes that can trigger a retry
 * @returns Configured retry request function
```

Reply:
Updated, thanks.